3.1.1 discriminator improvements (#2618)
OAI/OpenAPI-Specification: versions/3.1.1.md

Field Name | Type | Description
---|:---:|---

> <a name="schemaDiscriminator"></a>discriminator | [Discriminator
> Object](#discriminatorObject) | Adds support for polymorphism. The discriminator is used to determine which of a set of schemas a payload is expected to satisfy. See [Composition and Inheritance](#schemaComposition) for more details.

Reviewer: ```suggestion
<a name="schemaDiscriminator"></a>discriminator | [Discriminator Object](#discriminatorObject) | Adds support for polymorphism. The discriminator is used to determine which of a set of schemas a payload is expected to satisfy. See [Composition and Inheritance](#schemaComposition) for more details.
```
Remove unnecessary line-break within "Discriminator Object"

`allOf` takes an array of object definitions that are validated *independently* but together compose a single object.

While composition offers model extensibility, it does not imply a hierarchy between the models.
To support polymorphism, the OpenAPI Specification adds the `discriminator` keyword.

> When used, the `discriminator` will indicate the name of the property that hints which schema definition is expected to validate the structure of the model.

Reviewer 2: ```suggestion
When used, the `discriminator` indicates the name of the property that hints which schema definition is expected to validate the structure of the model.
```

#### <a name="discriminatorObject"></a>Discriminator Object

> When request bodies or response payloads may be one of a number of different schemas, a `discriminator` object can be used to aid in serialization, deserialization, and validation. The `discriminator` keyword is used to inform the consumer of the document which of the alternatives is expected or preferred.

Reviewer 2: I find the phrase "expected or preferred" strange.  Also, I'm not sure how "consumer" is defined in this sentence but it can be used by both the consumer and server.

```suggestion
When request bodies or response payloads may be one of a number of different schemas, a `discriminator` object gives a hit about the expected schema of the document. It can be used to aid in serialization, deserialization, and validation.
```
Also, why interchange "keyword" and "object"?

Author: > why interchange "keyword" and "object"?

"Keyword" refers to the key and "object" refers to the value. The `discriminator` keyword has an object as its value.

`discriminator` uses a schema's "name" to automatically map a property value to
a schema. The schema's "name" is the property name used when declaring the
schema as a component in an OpenAPI document. For example, the name of the
schema at `#/components/schemas/Cat` is "Cat". Therefore, when using
`discriminator`, _inline_ schemas will not be considered because they don't have
a "name".

Schema names are scoped to the OpenAPI document they are defined in. That means
if OpenAPI document "A" references a schema in OpenAPI document "B", the schema

> names in "A" aren't considered when evaluating the referenced schema in "B".

Reviewer 3: This is the main part that's controversial.  In Swagger 2.0, the references were strongly implied to be scoped to document "A", and many tools implement that behavior.  There are other component name connections that have similar ambiguities and we should address them all at once.  (FWIW, I strongly agree with what you've written here, but it would be a de-facto breaking change for some and we recently decided not to do "clarifications" that are de-facto breaking changes).

  - $ref: '#/components/schemas/Lizard'
```

> which means the payload _MUST_, by validation, match exactly one of the schemas described by `Cat`, `Dog`, or `Lizard`. Evaluating a `oneOf` can be a costly operation, so `discriminator` MAY be used as a "hint" to improve the efficiency of selection of the matching schema. The `discriminator` keyword cannot change the validation result of the `oneOf`, it can only help make the evaluation more efficient and provide better error messaging. We can specify the exact field that tells us which schema is expected to match the instance:

Reviewer 4: I'm a little hung up on this paragraph's use of "evaluation" -- does that mean "validation" and if so will that be clear to (knowledgable) readers? Also, it's not clear to mean that the discriminator makes _validation_ of a `oneOf` schema more efficient -- you still have to check _all_ of the schemas to know that exactly one matches. The value of the discriminator is that it can make serialization/deserialization more efficient.

Here's a revision that I think is clearer:
```suggestion
which means the payload _MUST_, by validation, match exactly one of the schemas described by `Cat`, `Dog`, or `Lizard`. Deserialization of a `oneOf` can be a costly operation, as it requires determining which schema matches the payload and thus should be used in deserialization. This problem also exists for `anyOf` schemas. A `discriminator` MAY be used as a "hint" to improve the efficiency of selection of the matching schema. The `discriminator` keyword cannot change the validation result of the `oneOf`, it can only help make the deserialization more efficient and provide better error messaging. We can specify the exact field that tells us which schema is expected to match the instance:
```

Author: > I'm a little hung up on this paragraph's use of "evaluation" -- does that mean "validation"

Generally, in JSON Schema land, we use the term "evaluation" rather than "validation" to acknowledge that there are multiple ways to evaluate a schema and not all of them are validation. Other types of evaluations include code generation, documentation generation, form generation, etc. However, I think in this case we're really just talking about validation, so it might be better to just say "validation".

> it's not clear to mean that the discriminator makes validation of a `oneOf` schema more efficient -- you still have to check all of the schemas to know that exactly one matches.

In most cases, you need to validate every option. The one case where you can short circuit is when you find two matching schemas. Once you find the second passing schema, you know validation will fail and you don't need to check the rest. `disciminator` can help you get there more quickly by checking the one you expect to fail first.

* Pass
* Fail
* Fail
* Fail
* Pass <- discriminator

Assuming the schemas will be evaluated in order, without a discriminator, you would have to check every schema. With a disciminator, you could check that one first and then when evaluating the rest in order, you would find the extra match right away and could skip evaluating the rest of the schemas.

It requires a bit of luck and a very specific scenario, but it's possible. The only reason I included this line was to keep the spirit of the original text, but I'm happy to take it out. It's technically correct, but the possibility of optimization is so small and so rare, it's not really worth mentioning. `anyOf` could provide a level of optimization worth mentioning, but no one uses it, which is why I used `oneOf` instead.

Reviewer 4: > With a disciminator, you could check that one first and then when evaluating the rest in order, you would find the extra match right away and could skip evaluating the rest of the schemas.

Ah, great point. I did not think of this, but now I see how a discriminator can also make validation more efficient.

  - $ref: '#/components/schemas/Lizard'
```

> which means the payload _MUST_, by validation, match exactly one of the schemas described by `Cat`, `Dog`, or `Lizard`. Evaluating a `oneOf` can be a costly operation, so `discriminator` MAY be used as a "hint" to improve the efficiency of selection of the matching schema. The `discriminator` keyword cannot change the validation result of the `oneOf`, it can only help make the evaluation more efficient and provide better error messaging. We can then describe exactly which field tells us which schema is expected to match the instance:

Reviewer 5: The description of the PR:
> The discriminator keyword can't be a hint and a shortcut.

but here in the spec I see:
> `discriminator` MAY be used as a "hint" to improve the efficiency of selection of the matching schema

Probably it's more about the PR description, but I think it still worth pointing out 😅

Author: I'm not following what you're pointing out. My goal with this change is to make it clear that `discriminator` doesn't change the validation behavior of the `oneOf`/`anyOf` it applies to. At best, it can be used internally by the validator to improve validation performance. For example, if you have a discriminated `anyOf` and the schema you expect to pass is the fifth schema listed, then `discriminator` can be used to tell the implementation to check the fifth schema first deferring the first, second, third, and fourth schemas that aren't expected to pass. But, if that fifth schema fails validation, the validator still needs to check the rest of the schemas to look for a match. So, the validation isn't shortcutted. You still have to do the same validation checks as you would if `discriminator` wasn't present. `discriminator` allows the validator to _prioritize_ the evaluation of a schema that is expected to pass, it doesn't _eliminate_ the other schemas from consideration.

Reviewer 5: Yup-yup, I see, and since `discriminator` MAY be used as a hint, description of the PR should be changed from:
> The discriminator keyword can't be a hint and a shortcut.
to maybe smth like this:
> The discriminator keyword can't be a shortcut, only a hint.

Reviewer 2: ```suggestion
which means the payload _MUST_, by validation, match exactly one of the schemas described by `Cat`, `Dog`, or `Lizard`. Evaluating a `oneOf` can be a costly operation, so `discriminator` MAY be used as a "hint" to improve the efficiency of selection of the matching schema. The `discriminator` keyword cannot change the validation result of the `oneOf`, it can only help make the evaluation more efficient and provide better error messaging. We can specify the exact field that tells us which schema is expected to match the instance:
```

>   - $ref: '#/components/schemas/Lizard'

Author: Done. Thanks.

```

> which means the payload _MUST_, by validation, match exactly one of the schemas described by `Cat`, `Dog`, or `Lizard`. Evaluating a `oneOf` can be a costly operation, so `discriminator` MAY be used as a "hint" to improve the efficiency of selection of the matching schema. The `discriminator` keyword can not change the validation result of the `oneOf`, it can only help make the evaluation more efficient and provide better error messaging. We can then describe exactly which field tells us which schema is expected to match the instance:

Reviewer: ```suggestion
which means the payload _MUST_, by validation, match exactly one of the schemas described by `Cat`, `Dog`, or `Lizard`. Evaluating a `oneOf` can be a costly operation, so `discriminator` MAY be used as a "hint" to improve the efficiency of selection of the matching schema. The `discriminator` keyword cannot change the validation result of the `oneOf`, it can only help make the evaluation more efficient and provide better error messaging. We can then describe exactly which field tells us which schema is expected to match the instance:
```

> which means the payload _MUST_, by validation, match exactly one of the schemas described by `Cat`, `Dog`, or `Lizard`. Evaluating a `oneOf` can be a costly operation, so `discriminator` MAY be used as a "hint" to improve the efficiency of validation and selection of the matching schema. The `discriminator` keyword can not change the validation result of the `oneOf`, it can only help make the evaluation more efficient and provide better error messaging. We can then describe exactly which field tells us which schema is expected to match the instance:

Reviewer 6: ```suggestion
which means the payload _MUST_, by validation, match exactly one of the schemas described by `Cat`, `Dog`, or `Lizard`. Evaluating a `oneOf` can be a costly operation, so `discriminator` MAY be used as a "hint" to improve the efficiency of selection of the matching schema. The `discriminator` keyword can not change the validation result of the `oneOf`, it can only help make the evaluation more efficient and provide better error messaging. We can then describe exactly which field tells us which schema is expected to match the instance:
```
This needs some more discussion...  @jdesrosiers @mkistler

Reviewer 4: This looks good to me, and I made a small suggestion above to preserve the statement about the discriminator property being required.

}
```

> Will indicate that the `Cat` schema is the alternative that is expected to match this payload.

Reviewer 2: ```suggestion
Will indicate that the `Cat` schema is expected to match this payload.
```

      monster: 'https://gigantic-server.com/schemas/Monster/schema.json'
```

> Here the discriminator _value_ of `dog` will map to the schema `#/components/schemas/Dog`, rather than the default (implicit) `#/components/schemas/dog`.  If the discriminator _value_ does not match an implicit or explicit mapping, no schema can be determined and validation SHOULD fail. Mapping keys MUST be string values, but tooling MAY convert response values to strings for comparison.

Reviewer 7: Doesn't this line now contradict the constraint added a few lines above?

> The `discriminator` keyword can not change the validation result of the `oneOf`, it can only help make the evaluation more efficient and provide better error messaging.

Author: In this case the validation result of the `discriminator` keyword is "fail". The validation result of the `oneOf` keyword is not affected. The validation result of the schema as whole is affected by this behavior, but not the validation result of the `oneOf` keyword.

In #2621, I remove this constraint because it doesn't seem necessary to cause validation to fail for a keyword that doesn't otherwise affect the validation result. I couldn't make that change here to avoid backward incompatible changes since this is a patch release.

```yaml
components:
  schemas:
    MyResponseType:
      oneOf:
      - $ref: '#/components/schemas/Cat'
      - $ref: '#/components/schemas/Dog'

>       - $ref: '#/components/schemas/Lizard'

Reviewer 3: My understanding is that the `"allOf"` form specifically exists to avoid having to list the subschemas in a `"oneOf"` or `"anyOf"`, so this should not be changed.  I don't personally understand how this even works, and [I'm not the only one](https://github.com/OAI/OpenAPI-Specification/issues/3591).  But it's related to the controversial thing- if all schema names resolve to the Components Object of what we now call the ["entry document"](https://learn.openapis.org/glossary.html), you can just search that one list.  I guess?  Apparently we need to figure that out.

Author: The problem I was trying to solve here is that the example payloads never say which schema is being used to evaluate those payloads. Based on my understanding of the `allOf` usage, none of those schemas could produce the results the documentation says it does, so I introduced something that could.

At this point I can only assume that the "Pet" schema is what these payloads are intended to be evaluated against and the behavior described is correct in that case. So, I've reverted my change other than to specify that "Pet" is the schema these payloads are evaluated against. Let me know if that's the right thing to do. I can't see how that makes sense, so I wouldn't be surprised if it's still not correct. Let me know.

            type: boolean
```

> The `MyResponseType` schema will use the discriminator defined by the `Pet`

Reviewer 5: This usage of `discriminator` deep inside inheritance tree seems quite complicated to implement and even reason about 🤔 Like, why should we use `MyResponseType.oneOf` and not `Cat.allOf`.

Shouldn't we require having `oneOf`/`anyOf`/`allOf` on the same level? We already enforce it in json schema btw, see https://github.com/OAI/OpenAPI-Specification/pull/3137.

Author: > Shouldn't we require having `oneOf`/`anyOf`/`allOf` on the same level? We already enforce it in json schema btw, see https://github.com/OAI/OpenAPI-Specification/pull/3137.

Good point! It seems that change was incorrect and should be backed out.

Reviewer 5: Added a PR with the revert: https://github.com/OAI/OpenAPI-Specification/pull/3405

schema because it is part of the `Cat`, `Dog`, and `Lizard` schemas in the
`oneOf`. The behavior if not all schemas define a `discriminator` and they are

> not all the same is undefined.

Reviewer 3: This would need to be removed if the `oneOf` addition to the example is removed.
